assignment4/Q1.cpp

```diff
@@ -0,0 +1,126 @@
+#include <iostream>
+#include <vector>
+#include <set>
+
+using namespace std;
+//iterating two times - first to copy the board
+//not sure of its correctness
+//probably something is missing but I do not know yet
```

Comment:
Your solution is not necessarily a union-set solution. It is a good solution, as using union-set was just a suggestion. Any other valid solutions, or more original solutions such as yours, are really great! :)

If I understood correctly, your algorithm is:
1. Keep a set of all different islands indices.
2. For each cell:
  a. if it has *no* top/left neighbors, declare it as a new island (with its new index).
  b. if it has *only one* neighbor, invalidate the cell's island index and merge it with its neighbor.
  c. If it has *both* top and left neighbors, merge the island with the top neighbor, and invalidate the left island (remove its index from the set).

After "invalidating" an island, it will not be counted in the total islands in the end. Also, using a set makes sure that invalidating an island more than once has the same effect as doing so once.

As I said, this algorithm is not a union-set algorithm - to be so, it should have one key feature which is missing here: a "rank" system that makes sure that the low-rank set is always merged into the high-rank set, to prevent "circular" merges; and also a system that makes sure that each set is pointing to its "root".
However, this assignment does not present a general case of a union set, but a very special case - a 2d array of tiles, where a tile can be merged only with its neighbor.
You chose to always merge the left cell with the top cell. And it worked, since they will never "change positions" and there would never be a circular merge in that case (Your algorithm will never merge index 1 into 2, and then merge 2 into 1).
Also, your method of "invalidating" islands (instead of making them point to the root island) works perfectly here, since you do not need to know if two tiles are from the same island, but merely to count them.
This is why your algorithm works, and your solution is good!

If you want to practice using the disjoint set algorithm anyway, you can refer to this wikipedia page:
https://en.wikipedia.org/wiki/Disjoint-set_data_structure
I'll be happy to help if you have further questions.